### PR TITLE
Add 60 second time control for Puyo duel

### PR DESCRIPTION
**Pull request by frostburn**
refs https://github.com/frostburn/panel-league/issues/128

### CODE REVIEW

File: lib/common/puyo/stepper.js [hunk 1]
```diff
@@ -121,7 +121,7 @@ class BasicStepper {
 
   canStep(state, events) {
     if (state.chainNumber) {
-      if (events.length) {
+      if (events.some(event => event.type === 'addPuyos')) {
```
Comment (RauliL) [inline lib/common/puyo/stepper.js]: `find` does the same thing. Not sure which one is faster though.

File: lib/common/puyo/stepper.js [hunk 2]
```diff
@@ -132,7 +132,7 @@ class BasicStepper {
   }
 
   canPlay(state, events) {
-    return (!events.length && !state.chainNumber);
+    return (!events.some(event => event.type === 'addPuyos') && !state.chainNumber);
```
Comment (RauliL) [inline lib/common/puyo/stepper.js]: Same here.

File: lib/common/puyo/stepper.js [hunk 3]
```diff
@@ -411,6 +413,9 @@ class DuelStepper extends BasicStepper {
   canStep(state, events) {
     const eventsForPlayer = this.distributeEvents(state, events);
 
+    if (events.some(event => event.type === 'termination')) {
```
Comment (RauliL) [inline lib/common/puyo/stepper.js]: And here.

File: lib/server/gamemode.js [hunk 4]
```diff
@@ -1,3 +1,28 @@
...
+  }
+
+  makeMove(event) {
+    if (event.type === 'addPuyos') {
+      super.makeMove(event.player);
+    }
+  }
+
+  step(engine) {
+    for (let player = 0; player < this.players.length; ++player) {
```
Comment (RauliL) [inline lib/server/gamemode.js]: Could be replaced with
```JavaScript
this.players.filter(player => !engine.callStepper('canPlay', { player })).forEach(player => super.makeMove(player));
```

File: lib/common/timecontrol.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  export() {
+    const data = super.export();
+
+    data.type = 'absolute';
+    data.maxTime = this.maxTime;
+    return data;
+  }
+}
+
+function TimeControlFactory(data) {
```
Comment (RauliL) [inline lib/common/timecontrol.js]: Functions use `thisTypeOfNamingConvention`.

File: lib/common/puyo/stepper.js (hunk 1 above)
Comment (frostburn) [inline lib/common/puyo/stepper.js]: It sure does.

**Comment by frostburn** [comment]
Did the refactorings.

**Comment by frostburn** [comment]
I heard that this was OK.

**Pull request merged**